src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -109527,5 +109527,764 @@ class Program
+        public void IEqualityComparer_SubSubInterfaceEqualsMethod_MaybeNullExpr_NonNullExpr_NoWarningWhenTrue()
+        {
+            var source = @"
+using System.Collections.Generic;
+
+#nullable enable
+
+interface MyEqualityComparer : IEqualityComparer<string?> { }
+interface MySubEqualityComparer : MyEqualityComparer { }
+
+class C
+{
+    static void M1(MySubEqualityComparer comparer, string? s)
+    {
+        if (comparer.Equals("""", s))
+        {
+            _ = s.ToString();
+        }
+        else
+        {
+            _ = s.ToString(); // 1
+        }
+    }
+}";
+
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (19,17): warning CS8602: Dereference of a possibly null reference.
+                    //             _ = s.ToString(); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(19, 17));
+        }
+
+        [Fact, WorkItem(35816, "https://github.com/dotnet/roslyn/issues/35816")]
+        public void OverrideEqualsMethod_MaybeNullExpr_NonNullExpr_NoWarningWhenTrue()
+        {
+            var source = @"
+#nullable enable
+class C
+{
+    public override bool Equals(object? other) => throw null!;
+    public override int GetHashCode() => throw null!;
+
+    static void M1(C? c1)
+    {
+        C c2 = new C();
+        if (c2.Equals(c1))
+        {
+            _ = c1.ToString();
+        }
+        else
+        {
+            _ = c1.ToString(); // 1
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (17,17): warning CS8602: Dereference of a possibly null reference.
+                    //             _ = c1.ToString(); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1").WithLocation(17, 17));
+        }
+
+        [Fact, WorkItem(35816, "https://github.com/dotnet/roslyn/issues/35816")]
+        public void NotImplementingEqualsMethod_MaybeNullExpr_NonNullExpr_Warns()
+        {
+            var source = @"
+#nullable enable
+class C
+{
+    public bool Equals(C? other) => throw null!;
+
+    static void M1(C? c1)
+    {
+        C c2 = new C();
+        if (c2.Equals(c1))
+        {
+            _ = c1.ToString(); // 1
+        }
+        else
+        {
+            _ = c1.ToString(); // 2
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (12,17): warning CS8602: Dereference of a possibly null reference.
+                    //             _ = c1.ToString(); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1").WithLocation(12, 17),
+                    // (16,17): warning CS8602: Dereference of a possibly null reference.
+                    //             _ = c1.ToString(); // 2
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1").WithLocation(16, 17));
+        }
```

Comment:
>} [](start = 8, length = 1)

Consider adding tests that don't rely on `string?` or `C? types, but rather also try `int?` or `TOpen?`.
Also, let's adding some tests for `notNull1` compared to `notNull2` (no learning, both remain not-null) and `maybeNull1` compared to `maybeNull2` (no learning, both remain maybe-null)